content/en/blog/fsm2/index.md

```diff
@@ -1,207 +1,357 @@
-  target: INITIAL
-states:                                              # (2)
-  INITIAL:                                           # (3)
-    transitions:                                     # (4)
-    - trigger:                                       # (5)
-        protocol: basic_message                      # (6)
-      sends:                                         # (7)
-      - data: Hello! I'm Hello-World bot.
-        protocol: basic_message                      # (8)
-      target: INITIAL                                # (9)
+As I said in [my previous blog
+post](https://findy-network.github.io/blog/2023/03/13/no-code-ssi-chatbots-part-i/),
+the idea of chatbots came quite naturally for our SSI development team. Think
+tanks or research labs are outstanding workplaces when you have something to chew.
+And how juicy topic the SSI has been, oh boy.
+
+FSM chatbots started as a thought experiment, but now, when I think of the
+invention, it has been quite clever in foreseeing genuine user needs. The
+same goes for technology. We desperately needed [better abstraction
+layers](http://findy-network.github.io/blog/2022/03/05/the-missing-network-layer-model/)
+over DID-shit.
+
+(*Do you know the nagging feeling that something is wrong? We are starting to
+ get our hands around it, thanks to these PoCs and technology spikes, but most
+ importantly, thru the use cases. My upcoming blog post will be all about
+ that.*)
+
+{{< imgproc cover-layers.png Resize "2553x" >}}
+<em>Network Layers For SSI Are Clarifying Them Selves</em>
+{{< /imgproc >}}
+
+Implementing the FSM chatbots with our protocol-agnostic API has been
+eye-opening. With our API, we have solved most of the complexity problems and
+brought in an elegant communication layer that hides most of the horror of DID
+system, which, by to way, is not application development ready.
+
+The preceding drawing presents our ongoing layer architecture. It's (still)
+based on DIDComm, but because our protocol engine implements a
+technology-agnostic API, we can change the implementation to something better
+without disturbing the above layers. The subject of this blog post is *our FSM
+engine* (3rd layer from bottom) which offers even more abstraction and helps the
+application development with no-code chatbot language.
+
+## Concurrency \\(\ne\\) Parallelism
+
+When I started playing with Go’s channels and [CSP programming
+model](https://en.wikipedia.org/wiki/Communicating_sequential_processes), I
+learned how easy it was compared to other asynchronous or multithreading
+programming models. I have been using multithreading programming for almost the
+whole of my career and closely monitoring paradigms of multithreading
+programming since the 90s.
+
+{{< imgproc interrupt.jpg Resize "463x" >}}
+<em>Interrupt Based Scheduling -- Mark Siegesmund, in Embedded C Programming, 2014</em>
+{{< /imgproc >}}
+
+If I look at the phases I have gone thru, they follow these steps harshly:
```

Comment:
This is really interesting chapter, I would like to learn more about the different technologies, but I guess it is off-topic for this blog 😄 .